### ISSUE TITLE
About the Doubly stochastic regularization 

### ISSUE THREAD

**Issue by jbdel** [yunjey/show-attend-and-tell-tensorflow#15]
Hello, 
I'm a little surprise by this code : 
```
if self.alpha_c > 0:
            alphas = tf.transpose(tf.pack(alpha_list), (1, 0, 2))     # (N, T, L)
            alphas_all = tf.reduce_sum(alphas, 1)      # (N, L)
            alpha_reg = self.alpha_c * tf.reduce_sum((16./196 - alphas_all) ** 2)     
```
I dont understand the calcul to get alpha_reg. As i understand, alphas_all is the sum of alphas for every spatial location. In the paper, we want this to be close from 1, shouldnt then it be something like : 

```
alpha_reg = self.alpha_c * tf.reduce_sum(tf.abs(1-alphas_all))    
```
Thanks in advance for the clarification.

**Comment by yunjey** [comment]
First, ```(16/196 - alphas_all) ** 2``` and ```(1 - alphas_all)``` are same in terms of optimization. The doubly stochastic regularization is done to evenly focus attention on all parts of the images. Since this code has a maximum sequence length of 16 and a spatial size of 196, all values in alphas_all should be 16/196 to see all parts of the image evenly.

**Comment by arieling** [comment]
I think it's problematic. In this way, we will count all alphas but we should never let back propagate flow goes into LSTM after <END>. 
I think
`alpha_list.append(tf.transpose(tf.mul(tf.transpose(alpha,[1,0]), mask[:, t]),[1,0]))`
will make more sense.

**Comment by YuanEZhou** [comment]
I agree with  @arieling , I think we can also achieve this goal as  follows.
if self.alpha_c > 0:
            alphas = tf.transpose(tf.pack(alpha_list), (1, 0, 2))  ***tf.expand_dims(mask,axis=2)**   # (N, T, L)
            alphas_all = tf.reduce_sum(alphas, 1)      # (N, L)
            alpha_reg = self.alpha_c * tf.reduce_sum((1. - alphas_all) ** 2)

**Comment by arieling** [comment]
@YuanEZhou but still in my experiment, it doesn't change result a lot

**Comment by YuanEZhou** [comment]
Yeah, I had run the code, but the value of BLEU_1  is  0.662.  It can't achieve the level which the paper claimed. Do you have any suggestion to adjust the code?

**Comment by arieling** [comment]
@YuanEZhou data split is different.I think the difference is reasonable.
